File: src/main/kotlin/p0036_valid_sudoku/solution.kt
package p0036_valid_sudoku

class Solution {
    fun isValidSudoku(board: Array<CharArray>): Boolean {
        val hCheck = BooleanArray(9)
        val vCheck = BooleanArray(9)
        val bCheck = BooleanArray(9)
        repeat(9){ j ->
            val rowStart: Int = (j / 3) * 3
            val columnStart: Int = (j % 3) * 3

            repeat(9){ k ->

                // horizontal
                var code = board[j][k].code - 49
                if (code >= 0){
                    if (hCheck[code])
                        return false
                    hCheck[code] = true
                }

                // vertical
                code = board[k][j].code - 49
                if (code >= 0){
                    if (vCheck[code])
                        return false
                    vCheck[code] = true
                }

                // box
                code = board[rowStart + k/3][columnStart + k%3].code - 49
                if (code >= 0){
                    if (bCheck[code])
                        return false
                    bCheck[code] = true
                }
            }

            repeat(9){
                hCheck[it] = false
                vCheck[it] = false
                bCheck[it] = false
            }
        }

        return true
    }
}

File: src/main/kotlin/p0036_valid_sudoku/solution.cpp
#include <math.h>
#include <memory.h>
#include <iostream>
#include <algorithm>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        
        bool hCheck[10];
        bool vCheck[10];
        bool bCheck[10];

        for (size_t i = 0; i < 9; i++){
            int rowStart = (i/3) * 3;
            int columnStart = (i%3) * 3;

            for (size_t j = 0; j < 9; j++){
                // horizontal
                char s = board[i][j];
                if (s != 46){
                    int num = (int)s - 48;
                    if (hCheck[num])
                        return false;
                    hCheck[num] = true;
                }

                // vertical
                s = board[j][i];
                if (s != 46){
                    int num = (int)s - 48;
                    if (vCheck[num])
                        return false;
                    vCheck[num] = true;
                }

                // box
                s = board[rowStart + j/3][columnStart + j%3];
                if (s != 46){
                    int num = (int)s - 48;
                    if (bCheck[num])
                        return false;
                    bCheck[num] = true;
                }
            }

            for (size_t j = 0; j < 10; j++){
                hCheck[j] = false;
                vCheck[j] = false;
                bCheck[j] = false;
            }
        }
        
        return true;
    }
};

int main() {
    Solution* s = new Solution();

    int size;
    cin >> size;

    vector<int> nums;

    for (size_t i = 0; i < size; i++) {
        int num;
        cin >> num;
        nums.push_back(num);
    }

    return 0;
}
